crates/core/src/diff/fragment/merge.rs

```diff
@@ -0,0 +1,451 @@
+                let old_id = *id < 0;
+                let abs_id = id.abs().to_string();
+
+                let component = if old_id {
+                    old_components
+                        .get(&abs_id)
+                        .ok_or(MergeError::MissingComponent(*id))?
+                } else {
+                    new_components
+                        .get(&abs_id)
+                        .ok_or(MergeError::MissingComponent(*id))?
+                };
+
+                let mut comp = component.clone();
+                comp.resolve_cids(old_components, new_components)?;
+
+                let ComponentStatics::Statics(s) = comp.statics else {
+                    return Err(MergeError::UnresolvedComponent);
+                };
+
+                *self = Child::Fragment(Fragment::Regular {
+                    statics: Some(Statics::Statics(s)),
+                    reply: None,
+                    children: comp.children,
+                });
+
+                Ok(())
+            }
+        }
+    }
+}
+
+impl Component {
+    fn resolve_cids(
+        &mut self,
+        old_components: &HashMap<String, Component>,
+        new_components: &HashMap<String, Component>,
+    ) -> Result<(), MergeError> {
+        for child in self.children.values_mut() {
+            child.resolve_cids(old_components, new_components)?;
+        }
+
+        match self.statics {
+            ComponentStatics::ComponentRef(id) => {
+                let old_id = id < 0;
+                let abs_id = id.abs().to_string();
+
+                let component = if old_id {
+                    old_components
+                        .get(&abs_id)
+                        .ok_or(MergeError::MissingComponent(id))?
+                } else {
+                    new_components
+                        .get(&abs_id)
+                        .ok_or(MergeError::MissingComponent(id))?
+                };
+
+                let mut comp = component.clone();
+                comp.resolve_cids(old_components, new_components)?;
+
+                // currently the spec states that components should
+                // be merged and resolved by copying statics from the source tree
+                // https://github.com/phoenixframework/phoenix_live_view/blob/93d242460f5222b1d89e54df56624bc96d53d659/assets/js/phoenix_live_view/rendered.js#L238
+                self.statics = comp.statics;
+
+                // TODO: validate this behavior in JS, it seems like
+                // it is a straight replacement as opposed to a merge.
+                // to validate have a deeply nested component with children [1: {..}, 2: {..}, 3: {..}] and a diff with
+                // children [1: {..}, 3: {..}] and see who's kids win
+                if self.children.is_empty() {
+                    self.children = comp.children;
+                }
+            }
+            _ => {
+                // raw statics are fine
+            }
+        };
+        Ok(())
+    }
+}
```

Comment:
The main difference between the old merging logic and new merging logic is here.